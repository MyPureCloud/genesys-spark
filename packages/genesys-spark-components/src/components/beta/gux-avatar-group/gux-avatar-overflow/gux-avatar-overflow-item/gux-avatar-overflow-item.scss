@use '~genesys-spark/dist/beta/tokens/flare/global.scss';
@use '~genesys-spark/dist/beta/tokens/flare/ui.scss';
@use '~genesys-spark/dist/scss/focus.scss';

::slotted(img) {
  inline-size: 100%;
  block-size: 100%;
  object-fit: cover;
}

button {
  all: unset;
  box-sizing: border-box;
  display: flex;
  gap: ui.$gse-ui-menu-option-gap;
  align-items: center;
  inline-size: 100%;
  min-block-size: ui.$gse-ui-menu-option-height;
  padding: ui.$gse-ui-menu-option-padding;
  font-family: ui.$gse-ui-menu-option-label-default-text-fontFamily;
  font-size: ui.$gse-ui-menu-option-label-default-text-fontSize;
  font-weight: ui.$gse-ui-menu-option-label-default-text-fontWeight;
  line-height: ui.$gse-ui-menu-option-label-default-text-lineHeight;
  color: ui.$gse-ui-menu-option-label-foregroundColor;
  word-wrap: break-word;
  cursor: pointer;
  outline: none;
  outline-offset: calc(#{ui.$gse-ui-menu-option-focus-border-width} * -1);
  background-color: ui.$gse-ui-menu-option-default-backgroundColor;
  border: none;

  &:focus-visible {
    outline: ui.$gse-ui-menu-option-focus-border-width
      ui.$gse-ui-menu-option-focus-border-style
      ui.$gse-ui-menu-option-focus-border-color;
    border-radius: global.$gse-semantic-focusOutline-sm-borderRadius;
  }

  &:hover {
    background: ui.$gse-ui-menu-option-hover-backgroundColor;
  }

  &:active {
    font-family: ui.$gse-ui-menu-option-label-active-text-fontFamily;
    font-size: ui.$gse-ui-menu-option-label-active-text-fontSize;
    font-weight: ui.$gse-ui-menu-option-label-active-text-fontWeight;
    line-height: ui.$gse-ui-menu-option-label-active-text-lineHeight;
    background: ui.$gse-ui-menu-option-selected-backgroundColor;
  }

  .gux-avatar {
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: center;
    inline-size: ui.$gse-ui-avatar-xsmall-content-size;
    block-size: ui.$gse-ui-avatar-xsmall-content-size;
    padding: 0;
    margin: 0;
    overflow: hidden;
    color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
    cursor: pointer;
    background: none;
    background-color: ui.$gse-ui-avatar-media-initialsBackground-default;
    border: none;
    border-radius: 50%;

    .gux-avatar-initials {
      font-family: ui.$gse-ui-avatar-small-initials-fontFamily;
      font-size: ui.$gse-ui-avatar-xsmall-initials-fontSize;
      font-weight: ui.$gse-ui-avatar-xsmall-initials-fontWeight;
      line-height: ui.$gse-ui-avatar-small-initials-lineHeight;
      text-transform: uppercase;
    }

    &.gux-accent-1 {
      color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent1;
    }

    &.gux-accent-2 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent2;
    }

    &.gux-accent-3 {
      color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent3;
    }

    &.gux-accent-4 {
      color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent4;
    }

    &.gux-accent-5 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent5;
    }

    &.gux-accent-6 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent6;
    }

    &.gux-accent-7 {
      color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent7;
    }

    &.gux-accent-8 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent8;
    }

    &.gux-accent-9 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent9;
    }

    &.gux-accent-10 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent10;
    }

    &.gux-accent-11 {
      color: ui.$gse-ui-avatar-media-initialsForeground-default;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent11;
    }

    &.gux-accent-12 {
      color: ui.$gse-ui-avatar-media-initialsForeground-inverse;
      background-color: ui.$gse-ui-avatar-media-initialsBackground-accent12;
    }

    &.gux-accent-inherit {
      color: inherit;
      background-color: inherit;
    }
  }
}
