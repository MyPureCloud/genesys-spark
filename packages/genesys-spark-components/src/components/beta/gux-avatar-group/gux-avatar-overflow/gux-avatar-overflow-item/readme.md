# gux-avatar-overflow-item-beta



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                 | Type                                                                                                                             | Default     |
| -------- | --------- | --------------------------- | -------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `accent` | `accent`  | Manually sets avatar accent | `"0" \| "1" \| "10" \| "11" \| "12" \| "2" \| "3" \| "4" \| "5" \| "6" \| "7" \| "8" \| "9" \| "auto" \| "default" \| "inherit"` | `'auto'`    |
| `name`   | `name`    |                             | `string`                                                                                                                         | `undefined` |


## Slots

| Slot      | Description   |
| --------- | ------------- |
| `"image"` | Avatar photo. |


## Dependencies

### Used by

 - [gux-avatar-group-beta](../..)

### Graph
```mermaid
graph TD;
  gux-avatar-group-beta --> gux-avatar-overflow-item-beta
  style gux-avatar-overflow-item-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
