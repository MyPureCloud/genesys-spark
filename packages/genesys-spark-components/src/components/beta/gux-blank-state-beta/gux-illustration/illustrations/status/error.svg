<svg viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><g clip-path="url(#clip0_65666_645)"><path d="M0 16C0 20.2435 1.68571 24.3131 4.68629 27.3137C7.68687 30.3143 11.7565 32 16 32C20.2435 32 24.3131 30.3143 27.3137 27.3137C30.3143 24.3131 32 20.2435 32 16C32 11.7565 30.3143 7.68687 27.3137 4.68629C24.3131 1.68571 20.2435 0 16 0C11.7565 0 7.68687 1.68571 4.68629 4.68629C1.68571 7.68687 0 11.7565 0 16ZM10.5063 19.8563C10.5125 19.8125 10.5187 19.7625 10.525 19.7125C10.5437 19.6187 10.5687 19.525 10.6062 19.4375C10.6812 19.2563 10.7875 19.0875 10.9375 18.9438C11.9187 17.9625 12.9 16.9813 13.875 16.0063C12.8938 15.025 11.9187 14.05 10.9375 13.0688C10.35 12.4813 10.35 11.5312 10.9375 10.95C11.2312 10.6562 11.6125 10.5125 12 10.5125C12.3875 10.5125 12.7688 10.6562 13.0625 10.95C14.0437 11.9312 15.025 12.9125 16 13.8875C16.9813 12.9062 17.9625 11.925 18.9375 10.95C19.2313 10.6562 19.6125 10.5125 20 10.5125C20.3875 10.5125 20.7687 10.6562 21.0625 10.95C21.65 11.5375 21.65 12.4875 21.0625 13.0688C20.0812 14.05 19.1063 15.025 18.125 16.0063C19.1063 16.9875 20.0875 17.9688 21.0625 18.9438C21.3563 19.2375 21.5 19.6187 21.5 20.0063C21.5 20.3938 21.3563 20.775 21.0625 21.0688C20.7687 21.3625 20.3875 21.5063 20 21.5063C19.6125 21.5063 19.2313 21.3625 18.9375 21.0688L16 18.1313C15.0188 19.1125 14.0437 20.0875 13.0625 21.0688C12.7688 21.3625 12.3875 21.5063 12 21.5063C11.6125 21.5063 11.2312 21.3625 10.9375 21.0688C10.7937 20.925 10.6812 20.75 10.6062 20.575C10.5687 20.4875 10.5437 20.3937 10.525 20.3C10.5125 20.2312 10.5 20.1562 10.5063 20.1625C10.5 20.0875 10.5 20.0875 10.5 20.0125C10.5 19.9188 10.5063 19.8937 10.5063 19.8687V19.8563Z"/><path d="M13.0625 10.9375C12.475 10.35 11.525 10.35 10.9438 10.9375C10.3625 11.525 10.3563 12.475 10.9438 13.0563L13.8813 15.9938L10.9438 18.9313C10.3563 19.5188 10.3563 20.4688 10.9438 21.05C11.5313 21.6313 12.4813 21.6375 13.0625 21.05L16 18.1125L18.9375 21.05C19.525 21.6375 20.475 21.6375 21.0563 21.05C21.6375 20.4625 21.6438 19.5125 21.0563 18.9313L18.1188 15.9938L21.0563 13.0563C21.6438 12.4688 21.6438 11.5188 21.0563 10.9375C20.4688 10.3563 19.5188 10.35 18.9375 10.9375L16 13.875L13.0625 10.9375Z"/></g><defs><clipPath><rect/></clipPath></defs></svg>