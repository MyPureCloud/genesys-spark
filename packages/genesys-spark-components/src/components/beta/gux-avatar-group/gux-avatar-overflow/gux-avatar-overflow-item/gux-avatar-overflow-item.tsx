import { Component, h, JSX, Prop, Element, Listen, Host } from '@stencil/core';
import { trackComponent } from '@utils/tracking/usage';
import { logWarn } from '@utils/error/log-error';
import { generateInitials } from '@utils/string/generate-initials';
import { overflowNavigation } from '../gux-avatar-overflow.service';
import { getAvatarAccentClass } from '@components/beta/gux-avatar/gux-avatar.service';
import { GuxAvatarAccent } from '@components/beta/gux-avatar/gux-avatar.types';

/**
 * @slot image - Avatar photo.
 */

@Component({
  styleUrl: 'gux-avatar-overflow-item.scss',
  tag: 'gux-avatar-overflow-item-beta',
  shadow: { delegatesFocus: true }
})
export class GuxAvatarOverflowItem {
  @Element()
  root: HTMLElement;

  @Prop()
  name: string;

  /**
   * Manually sets avatar accent
   */
  @Prop()
  accent: GuxAvatarAccent = 'auto';

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
  }

  componentDidLoad() {
    this.validatingInputs();
  }

  @Listen('keydown')
  onKeydown(event: KeyboardEvent): void {
    overflowNavigation(event, this.root);
  }

  private validatingInputs(): void {
    const avatarImage = this.root.querySelector('img');
    if (!this.name) {
      logWarn(this.root, 'Name prop is required');
    }

    if (avatarImage && !avatarImage.getAttribute('alt')) {
      logWarn(this.root, 'Alt attribute is required for slotted image.');
    }
  }

  render(): JSX.Element {
    return (
      <Host role="menuitem">
        <button type="button" aria-label={this.name} tabIndex={-1}>
          <span
            class={{
              'gux-avatar': true,
              [getAvatarAccentClass(this.accent, this.name)]: true
            }}
          >
            <slot name="image">
              <span class="gux-avatar-initials" aria-hidden="true">
                {generateInitials(this.name)}
              </span>
            </slot>
          </span>
          <span>{this.name}</span>
        </button>
      </Host>
    ) as JSX.Element;
  }
}
