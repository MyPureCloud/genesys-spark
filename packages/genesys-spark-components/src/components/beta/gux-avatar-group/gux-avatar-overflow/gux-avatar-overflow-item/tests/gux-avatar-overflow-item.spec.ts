jest.mock('../../../../../../utils/error/log-error', () => ({
  __esModule: true,
  logWarn: jest.fn()
}));

import { newSpecPage } from '@stencil/core/testing';
import { logWarn } from '../../../../../../utils/error/log-error';
import { GuxAvatarOverflowItem } from '../gux-avatar-overflow-item';

const components = [GuxAvatarOverflowItem];
describe('gux-avatar-overflow-item-beta', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should build', async () => {
    const page = await newSpecPage({
      components,
      html: `<gux-avatar-overflow-item-beta name="John Doe"></gux-avatar-overflow-item-beta>`
    });
    expect(page.rootInstance).toBeInstanceOf(GuxAvatarOverflowItem);
  });

  describe('validatingInputs', () => {
    it('should log warning when name is not provided', async () => {
      await newSpecPage({
        components,
        html: `<gux-avatar-overflow-item-beta></gux-avatar-overflow-item-beta>`
      });

      expect(logWarn).toHaveBeenCalled();
    });

    it('should log warning when image has no alt attribute', async () => {
      await newSpecPage({
        components,
        html: `
            <gux-avatar-overflow-item-beta name="John Doe">
              <img slot="image" src="test.jpg"/>
            </gux-avatar-overflow-item-beta>
          `
      });

      expect(logWarn).toHaveBeenCalled();
    });
  });

  describe('Render', () => {
    it('should render with name', async () => {
      const page = await newSpecPage({
        components,
        html: `<gux-avatar-overflow-item-beta name="John Doe"></gux-avatar-overflow-item-beta>`
      });

      expect(page.root).toMatchSnapshot();
    });

    it('should render with slotted image', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <gux-avatar-overflow-item-beta name="John Doe">
            <img slot="image" src="test.jpg" alt="John Doe"/>
          </gux-avatar-overflow-item-beta>
        `
      });

      expect(page.root).toMatchSnapshot();
    });

    it('should render with initials when no image is provided', async () => {
      const page = await newSpecPage({
        components,
        html: `<gux-avatar-overflow-item-beta name="John Doe"></gux-avatar-overflow-item-beta>`
      });

      const initialsElement = page.root.shadowRoot.querySelector(
        '.gux-avatar-initials'
      );
      expect(initialsElement.textContent).toBe('JD');
    });
  });

  describe('Events', () => {
    it('should handle keydown event', async () => {
      const page = await newSpecPage({
        components,
        html: `<gux-avatar-overflow-item-beta name="John Doe"></gux-avatar-overflow-item-beta>`
      });
      const instance = page.rootInstance;
      const event = new KeyboardEvent('keydown', { key: 'ArrowDown' });

      // Mock overflowNavigation function
      const overflowNavigationSpy = jest.fn();
      instance.onKeydown(event);

      expect(overflowNavigationSpy).not.toHaveBeenCalled();
    });
  });
});
