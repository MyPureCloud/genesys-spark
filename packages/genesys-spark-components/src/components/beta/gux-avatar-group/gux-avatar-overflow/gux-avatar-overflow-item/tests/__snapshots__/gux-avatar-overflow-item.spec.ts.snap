// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-avatar-overflow-item-beta Render should render with name 1`] = `
<gux-avatar-overflow-item-beta name="John Doe" role="menuitem">
  <template shadowrootmode="open" shadowrootdelegatesfocus>
    <button aria-label="John Doe" tabindex="-1" type="button">
      <span class="gux-accent-3 gux-avatar">
        <slot name="image">
          <span aria-hidden="true" class="gux-avatar-initials">
            JD
          </span>
        </slot>
      </span>
      <span>
        John Doe
      </span>
    </button>
  </template>
</gux-avatar-overflow-item-beta>
`;

exports[`gux-avatar-overflow-item-beta Render should render with slotted image 1`] = `
<gux-avatar-overflow-item-beta name="John Doe" role="menuitem">
  <template shadowrootmode="open" shadowrootdelegatesfocus>
    <button aria-label="John Doe" tabindex="-1" type="button">
      <span class="gux-accent-3 gux-avatar">
        <slot name="image">
          <span aria-hidden="true" class="gux-avatar-initials">
            JD
          </span>
        </slot>
      </span>
      <span>
        John Doe
      </span>
    </button>
  </template>
  <img alt="John Doe" slot="image" src="test.jpg">
</gux-avatar-overflow-item-beta>
`;
